pkg/controllers/user/pipeline/controller/pipelineexecution/registrycertsyncer.go

```diff
@@ -0,0 +1,180 @@
+
+const (
+	checkCertRotateInterval = 12 * time.Hour
+	//rotateThreshold set the rotation threshold so it will rotate at 90% of the total lifetime of the certificates.
+	rotateThreshold = 0.9
+	maxRetry        = 60
+)
+
+type RegistryCertSyncer struct {
+	clusterName string
+
+	pods                    v1.PodInterface
+	podLister               v1.PodLister
+	secrets                 v1.SecretInterface
+	secretLister            v1.SecretLister
+	managementSecretLister  v1.SecretLister
+	namespaceLister         v1.NamespaceLister
+	pipelineExecutionLister v3.PipelineExecutionLister
+	pipelineSettingLister   v3.PipelineSettingLister
+}
+
+func (s *RegistryCertSyncer) sync(ctx context.Context, syncInterval time.Duration) {
+	for range ticker.Context(ctx, checkCertRotateInterval) {
+		s.checkAndRotateCerts(ctx)
+	}
+}
+
+func (s *RegistryCertSyncer) checkAndRotateCerts(ctx context.Context) {
+	labelsSearchSet := labels.Set{utils.PipelineNamespaceLabel: "true"}
+	namespaces, err := s.namespaceLister.List("", labels.SelectorFromSet(labelsSearchSet))
+	if err != nil {
+		logrus.Error(err)
+		return
+	}
+	for _, ns := range namespaces {
+		secret, err := s.secretLister.Get(ns.Name, utils.RegistryCrtSecretName)
+		if apierrors.IsNotFound(err) {
+			continue
+		} else if err != nil {
+			logrus.Error(err)
+			return
+		}
+		if !s.shouldRotate(secret.Data[utils.RegistryCrt]) {
+			continue
+		}
+		projectID := getProjectID(ns)
+		if projectID != "" {
+			go func() {
+				if err := s.rotateCerts(ctx, projectID); err != nil {
+					logrus.Error(err)
+				}
+			}()
+		}
+	}
+
+}
+
+func (s *RegistryCertSyncer) shouldRotate(certPEM []byte) bool {
+	currentCerts, err := cert.ParseCertsPEM(certPEM)
+	if err != nil || len(currentCerts) < 1 {
+		return true
+	}
+	currentCert := currentCerts[0]
+	totalDuration := currentCert.NotAfter.Sub(currentCert.NotBefore)
+	thresholdTime := currentCert.NotBefore.Add(time.Duration(float64(totalDuration) * rotateThreshold))
+	if time.Now().Before(thresholdTime) {
+		return false
+	}
+	return true
+}
+
+func (s *RegistryCertSyncer) rotateCerts(ctx context.Context, projectID string) error {
+	logrus.Debugf("rotating registry certs for %s project", projectID)
+
+	clusterID, projectID := ref.Parse(projectID)
+	//do certificate rotation when no pipeline execution is running
+	startTime := time.Now()
+	set := labels.Set(map[string]string{utils.PipelineFinishLabel: "false"})
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	defer cancelFunc()
+	for range ticker.Context(cancelCtx, time.Minute) {
+		runningExecutions, err := s.pipelineExecutionLister.List(projectID, set.AsSelector())
+		if err != nil {
+			return err
+		}
+		if len(runningExecutions) == 0 {
+			break
+		}
+		if time.Now().After(startTime.Add(maxRetry * time.Minute)) {
+			return errors.New("time out waiting all executions to finish")
+		}
+	}
+
+	//generate domain cert & key if they do not exist
```

Comment:
this method is a bit too long. Lets extract lines 116-162 to generateCert method (the name is up to you)